### ISSUE TITLE
The Big Rewrite

### ISSUE THREAD

**Issue by Changaco** [liberapay/liberapay.com#2330]
This issue is for the rewrite of a significant portion of Liberapay's source code that I first mentioned in <https://github.com/liberapay/salon/issues/541#issuecomment-1420578433>. The basic idea is to switch from SQL to a new database technology built from scratch in Python. The plan is to close all the technical issues related to the database (namely #245, #762, #980, #1113, #1312, #1595, #1701, #1727, #1736, #1962 and #2010) and take advantage of the migration to fix inaccuracies and inconsistencies in the data.

Sadly the work is still very far from complete. It's a big change and was always going to take a while, but there's also the problem that too much of my time and energy is consumed by other things.

**Comment by Changaco** [comment]
> Why?

Because it's painfully slow and difficult to build something correctly when you don't have the right tools. I've been working on this Python+PostgreSQL code base for a decade now, and I've come to the conclusion that SQL isn't the right tool for this job. The work I've done so far on this issue has confirmed that the Liberapay source code can be significantly simplified and improved by replacing SQL queries with Python code.

The new Python module needed to make this work will be significantly simpler than an SQL database, because its job will be simpler. It won't need a query parser and planner, for example. It will fulfill two basic needs: consensus and storage. The consensus part is an implementation of [Raft](https://en.wikipedia.org/wiki/Raft_\(algorithm\)) that I've already partly written, from scratch based on [the Raft paper](https://raft.github.io/raft.pdf). The storage part will be an implementation of well-known data structures. Efficiency will be difficult to achieve with Python, but it might not be necessary to fully optimize the storage layer right from the start.

A distinctive feature of the new API is its intuitive and automatic way of handling the fact that data changes over time. An SQL database can of course store multiple revisions of the same information, but it doesn't know if it's doing that, so for example it can't make it easy to manage how many revisions should be kept or whether to delta-compress them to save space.

**Comment by boehs** [comment]
This feels domain specific, which is fair. Efficiency and time for development were my main concerns. Would this replace everything, even stuff like user accounts?

**Comment by Changaco** [comment]
I wouldn't say that there's anything “domain specific” in this.

SQL will be completely replaced. Liberapay will no longer require or use a PostgreSQL database.

**Comment by boehs** [comment]
Well, I wish you luck in this endeavor

**Comment by Changaco** [comment]
(Cloudflare is continuing to push SQL-based solutions: [Zero-latency SQLite storage in every Durable Object](https://blog.cloudflare.com/sqlite-in-durable-objects/). I wonder if they've already realized that it's counterproductive to use SQL this way and are already aiming to remove that layer of abstraction in the future.)

**Comment by tif-calin** [comment]
> by replacing SQL queries with Python code

Have you looked into ORM options?

**Comment by Changaco** [comment]
> Have you looked into ORM options?

Fully solving the problems I've identified requires removing SQL entirely. Hiding it under an ORM like [SQLAlchemy](https://www.sqlalchemy.org/) would add complexity and further degrade performance.

**Comment by galuszkak** [comment]
Are we talking about something like ZODB? https://zodb.org/en/latest/index.html ?

**Comment by Changaco** [comment]
Yes and no. The new module is both similar to and quite different from ZODB. It's currently named hahadb, for “high availability history aware database”. ZODB has neither of those qualities.

**Comment by Changaco** [comment]
(@galuszkak I actually remember your username from the old Gittip/Gratipay days. Welcome “back”.)

**Comment by Changaco** [comment]
Cloudflare has been [experimenting](https://blog.cloudflare.com/meerkat-introduction/) with [QuePaxa](https://github.com/dedis/quepaxa), a consensus algorithm published in 2023. It looks like it could advantageously replace Raft in hahadb.
